### PR TITLE
[WIP] Initial httpd image

### PR DESCRIPTION
**Pull request by mnagy**
This is my first stab at the httpd image that serves static html pages. I've not yet looked into how to use this for our perl and php images, I will do that tomorrow. I've basically took conf files as they were in httpd24 SCL and copied them into a contrib directory and modified them. This allows us to run the container as apache user.

I didn't do any tests yet, not sure if we want to have this as a standalone image as well.

cc @mfojtik @bparees @jhadvig: You guys might want to see this.

### CODE REVIEW

File: 2.4/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+# Environment:
+#  None
+
+# Image metadata
+ENV HTTPD_VERSION           2.4
+ENV IMAGE_DESCRIPTION       Apache httpd server, version 2.4
+ENV IMAGE_TAGS              httpd,httpd24
+ENV IMAGE_EXPOSE_SERVICES   8080:http
+
+EXPOSE 8080
```
Comment (mfojtik) [inline 2.4/Dockerfile:15]: any reason why this is 8080? i thought the standard http port is 80

File: 2.4/contrib/httpd/conf/httpd.conf [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+# in which case these default settings will be overridden for the
+# virtual host being defined.
+#
+
+#
+# ServerAdmin: Your address, where problems with the server should be
+# e-mailed.  This address appears on some server-generated pages, such
+# as error documents.  e.g. admin@your-domain.com
+#
+ServerAdmin root@localhost
```
Comment (mfojtik) [inline 2.4/contrib/httpd/conf/httpd.conf]: apache@localhost

Comment (mfojtik) [inline 2.4/contrib/httpd/conf/httpd.conf]: actually this should be configurable

**Comment by mfojtik** [comment]
This is STI enabled image and I miss .sti/bin/assemble and .sti/bin/run script (which will be the entrypoint)

File: 2.4/Dockerfile (hunk 1 above)
Comment (bparees) [inline 2.4/Dockerfile:15]: would require root to run on 80.

Comment (bparees) [inline 2.4/Dockerfile:15]: also we've standardized on 8080 for all our other images.

**Comment by mnagy** [comment]
Lot of changes. This is actually starting to look good. The scl enablement is pretty good. You only need to add a file to /opt/openshift/scl_enable.d/ and it'll get sourced in all cases, login shell or not, interactive or not, even in scripts, without you needing to do anything else.

This image can now be used as a base image for others. I already tried reworking the PHP image a bit to work with it: https://github.com/bparees/sti-php/issues/1

@mfojtik @bparees PTAL

**Comment by mnagy** [comment]
Also, no rhel7 yet, will add that later if there won't be any bad reviews :)

Also, this has a disadvantage that the downstream images need to COPY contrib and do other stuff that is normally handled by ONBUILD instructions that are in sti-base. The reason is that with ONBUILD instructions, the image can't be used with sti to serve static html pages.

**Comment by mnagy** [comment]
@mfojtik @bparees any comments?

File: 2.4/.sti/bin/assemble [hunk 3]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+
+set -e
+
+echo "---> Installing application source"
```
Comment (mfojtik) [inline 2.4/.sti/bin/assemble:5]: add " ..." suffix (we agreed to change it in all images, so I will be now enforcing it in all pull requests ;-)

File: 2.4/.sti/bin/run [hunk 4]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+source /opt/openshift/httpd_util.sh
```
Comment (mfojtik) [inline 2.4/.sti/bin/run:3]: this is not what RHCSL folks will want to see

File: 2.4/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+RUN rpmkeys --import file:///etc/pki/rpm-gpg/RPM-GPG-KEY-CentOS-7 && \
+    yum -y --setopt=tsflags=nodocs install https://www.softwarecollections.org/en/scls/rhscl/httpd24/epel-7-x86_64/download/rhscl-httpd24-epel-7-x86_64.noarch.rpm && \
+    yum -y --setopt=tsflags=nodocs install httpd24 && \
+    yum clean all && \
+    mkdir -p /opt/openshift/www/html && \
+    chown -R default.default /opt/openshift/www /var/log/httpd24 /opt/rh/httpd24/root/var/run/httpd
+
+# Make sure SCL is enabled for every interactive and/or login sessions.
+COPY scl_enable.sh /etc/profile.d/
```
Comment (mfojtik) [inline 2.4/Dockerfile:25]: is it worth to have 2 layers just to copy one file to multiple directories? can't you just symlink this?

File: 2.4/contrib/httpd_util.sh [hunk 6]
```diff
@@ -0,0 +1,14 @@
+# source
```
Comment (mfojtik) [inline 2.4/contrib/httpd_util.sh:1]: source what?

File: 2.4/.sti/bin/run (hunk 4 above)
Comment (mnagy) [inline 2.4/.sti/bin/run:3]: I am aware, but I don't know about any location that I should prefer :(

File: 2.4/contrib/httpd_util.sh [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+export HTTPD_SERVER_ADMIN=${HTTPD_SERVER_ADMIN:-default@localhost}
+
+function configure_httpd() {
+  echo "---> Configuring Apache"
```
Comment (mfojtik) [inline 2.4/contrib/httpd_util.sh:6]: add " ..." here as you want till this finish. Would be also useful to explain users what is actually being configured.

File: 2.4/contrib/httpd_util.sh [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+  for template in $(find /opt/openshift/httpd/ -name '*.template'); do
```
Comment (mfojtik) [inline 2.4/contrib/httpd_util.sh:7]: the /opt/openshift dir is unfortunate and will change (RHCSL wants to occupy the /usr/local/etc)

File: 2.4/contrib/httpd_util.sh (hunk 6 above)
Comment (mfojtik) [inline 2.4/contrib/httpd_util.sh:1]: I think this is pretty ugly. Requiring people that use this image as a base layer to 'source' some script is not what I think we should be doing.

The better approach would be to rename/replace the default ENTRYPOINT in the latter image (you can preserve it and call it from your ENTRYPOINT or you can just override it).

File: 2.4/scl_enable.sh [hunk 9]
```diff
@@ -0,0 +1,5 @@
+for i in /opt/openshift/scl_enable.d/*.sh ; do
```
Comment (mfojtik) [inline 2.4/scl_enable.sh:1]: this black magic should be done elsewhere, speak to RHCSL folks

File: 2.4/test/html-test-app/index.html [hunk 10]
```diff
@@ -0,0 +1,10 @@
+<!DOCTYPE html>
```
Comment (mfojtik) [inline 2.4/test/html-test-app/index.html:1]: Can we generate this default page and make it show something useful? Like the container environment or help? For example "how I replace this page with my page?"

File: 2.4/.sti/bin/run (hunk 4 above)
Comment (mfojtik) [inline 2.4/.sti/bin/run:3]: `/usr/local` was the last compromise. we should be switching to that.

**Comment by mfojtik** [comment]
You can copy the ONBUILD to this image if you want to use it as a 'builder' in future. That will force people to follow the 'standards' (iow. have the files in the 'known' locations'). I'm aware that this image can also be used as it is, but in that case the ONBUILD are ignored, right?

**Comment by mnagy** [comment]
No, they are not. We were talking about that actually. If you use ONBUILDs, sti build will break for some reason. An ideal case would really be to have a base image with ONBUILDs and have an image that would only serve static html pages. Which in reality would be more useful as an httpd image usage example.

**Comment by mfojtik** [comment]
@mnagy yeah that is in case that you want to 'build' something using this image. I think we should talk whether this will be a 'builder' image or just serve as 'base' for other builders, like PHP, Python OR 'static'.

**Comment by mnagy** [comment]
cc @notroj @hanzz

**Comment by mnagy** [comment]
@mfojtik I'd like this image to be a base for other images (ONBUILD instructions included) and have another one that serves static files. I wonder if that one could be included in this same repository? It'd be perfect mainly as an example of how to base other images on it.

**Comment by mnagy** [comment]
This is ancient, closing.

**Pull request closed without merging**